src/ray/protobuf/common.proto

```diff
@@ -0,0 +1,123 @@
+
+// Type of a task.
+enum TaskType {
+  // Normal task.
+  NORMAL_TASK = 0;
+  // Actor creation task.
+  ACTOR_CREATION_TASK = 1;
+  // Actor task.
+  ACTOR_TASK = 2;
+}
+
+/// The task specification encapsulates all immutable information about the
+/// task. These fields are determined at submission time, converse to the
+/// `TaskExecutionSpec` may change at execution time.
+message TaskSpec {
+  // Type of this task.
+  TaskType type = 1;
+  // Language of this task.
+  Language language = 2;
+  // Function descriptor of this task, which is a list of strings that can
+  // uniquely describe the function to execute.
+  // For a Python function, it should be: [module_name, class_name, function_name]
+  // For a Java function, it should be: [class_name, method_name, type_descriptor]
+  repeated bytes function_descriptor = 3;
+  // ID of the job that this task belongs to.
+  bytes job_id = 4;
+  // Task ID of the task.
+  bytes task_id = 5;
+  // Task ID of the parent task.
+  bytes parent_task_id = 6;
+  // A count of the number of tasks submitted by the parent task before this one.
+  uint64 parent_counter = 7;
+  // Task arguments.
+  repeated TaskArg args = 8;
+  // Number of return objects.
+  uint64 num_returns = 9;
+  // Quantities of the different resources required by this task.
+  map<string, double> required_resources = 10;
+  // The resources required for placing this task on a node. If this is empty,
+  // then the placement resources are equal to the required_resources.
+  map<string, double> required_placement_resources = 11;
+  // Task specification for an actor creation task.
+  // This field is only valid when `type == ACTOR_CREATION_TASK`.
+  ActorCreationTaskSpec actor_creation_task_spec = 14;
+  // Task specification for an actor task.
+  // This field is only valid when `type == ACTOR_TASK`.
+  ActorTaskSpec actor_task_spec = 15;
+}
+
+// Argument in the task.
+message TaskArg {
+  // Object IDs for pass-by-reference arguments. Normally there is only one
+  // object ID in this list which represents the object that is being passed.
+  // However to support reducers in a MapReduce workload, we also support
+  // passing multiple object IDs for each argument.
+  repeated bytes object_ids = 1;
+  // Data for pass-by-value arguments.
+  bytes data = 2;
+}
+
+// Task spec of an actor creation task.
+message ActorCreationTaskSpec {
+  // ID of the actor that will be created by this task.
+  bytes actor_id = 2;
+  // The max number of times this actor should be recontructed.
+  // If this number of 0 or negative, the actor won't be reconstructed on failure.
+  uint64 max_actor_reconstructions = 3;
+  // The dynamic options used in the worker command when starting the worker process for
+  // an actor creation task. If the list isn't empty, the options will be used to replace
+  // the placeholder strings (`RAY_WORKER_OPTION_0`, `RAY_WORKER_OPTION_1`, etc) in the
+  // worker command.
+  repeated string dynamic_worker_options = 4;
```

Comment:
I know that this is not new in this PR, but can you explain what the `dynamic_worker_options` are?

Reply:
This option allows us to specify dynamic arguments of the `worker command`. In particular, this enables specifying JVM memory limit for an actor. Because the JVM options have to be set before starting the java process, and can't be set at runtime.